relay-server/src/services/store.rs

```diff
@@ -244,16 +247,28 @@ impl StoreService {
                     item,
                 )?,
                 ItemType::ReplayRecording => {
-                    self.produce_replay_recording(event_id, scoping, item, start_time, retention)?
+                    if item.replay_combined_payload() {
+                        replay_recording = Some(item);
+                    } else {
+                        self.produce_replay_recording(
+                            event_id, scoping, item, start_time, retention,
+                        )?;
+                    }
+                }
+                ItemType::ReplayEvent => {
+                    if item.replay_combined_payload() {
+                        replay_event = Some(item);
+                    } else {
+                        self.produce_replay_event(
+                            event_id.ok_or(StoreError::NoEventId)?,
+                            scoping.organization_id,
+                            scoping.project_id,
+                            start_time,
+                            retention,
+                            item,
+                        )?;
```

Comment:
I'll preface this by saying this is a personal opinion.  I think this section needs to be refactored.  If we add the missing states from my previous comment we'll have duplicated producer code.  We'll also be producing effects in two different locations in the function.  Over the long term I think this is more error prone than doing everything close together.

My suggestion is: aggregate the state and evaluate that state in one place.  In psuedo code it looks like this:

```rust
fn store_envelope() {
    for ... {
        match ... {
            ItemType::ReplayEvent => {
                replay_event = Some(item);
            }
            ItemType::ReplayRecording => {
                replay_recording = Some(item);
            }
        }
    }

    if item.replay_combined_payload() {
        // Handles all the cases for combined types.
        self.produce_replay_messages(replay, replay_recording)
    } else {
        // Handles all the cases for separate types.
        self.produce_replay_messages(replay, None)
        self.produce_replay_messages(None, replay_recording)
    }
}

fn produce_replay_messages(x, y) {
    match (x, y) => {
        (Some(x), Some(y)) => {
            // produce combined
        },
        (Some(x), None) => {
            // produce x
        },
        (None, Some(y)) => {
            // produce y
        }
        _ => {
            // do nothing
        }
    }
}
```

Reply:
thank you! will also change it as discussed so that we're still producing the replay event to the snuba topic directly from here as well.